feat: update run0edit to v0.4.3 w/ support for immutable flag

HastD opened the pull request:
The new version of [`run0edit`](https://github.com/HastD/run0edit) checks for the immutable flag on the file to be edited (or the directory if the file doesn't exist), and if so, asks the user whether to temporarily remove it to edit/create the file. This makes editing immutable files less error-prone, as the immutable flag is only removed for the brief window necessary to overwrite the file contents and reapplied immediately afterward; the script also compares the file contents after the immutable flag has been reapplied to ensure it was not modified by another program during that window.

There are also miscellaneous improvements to the script's argument handling, which now parses arguments in a more conventional way and accepts both `--help` and `--version` arguments, and the inner privileged script has been refactored to reduce code duplication.

HastD committed:
feat: update run0edit to v0.4.0 w/ support for immutable flag

The new version of `run0edit` checks for the immutable flag on the file
to be edited, and if so, asks the user whether to temporarily remove it
to edit the file. This makes editing immutable files less error-prone,
as the immutable flag is only removed for the brief window necessary to
overwrite the file contents and reapplied immediately afterward; the
script also compares the file contents after the immutable flag has been
reapplied to ensure it was not modified by another program during that
window.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
fix: handle case where lsattr gives error

ShadowSlayer1441 commented on files/system/usr/bin/run0edit:
```diff
@@ -71,21 +115,43 @@ else
-directory=$(dirname "$filename")
-trunc_filename=$(basename "$filename" | head -c 64)
-temp_filename=$(mktemp --tmpdir "${trunc_filename}.XXXXXXXXXX")
+directory=$(dirname -- "$filename")
```
Should probably check if the file's directory is immutable, before trying to create temp file.

HastD replied:
Empty temp files are cleaned up at the end of the script regardless, so I don't think this makes a difference.

ShadowSlayer1441 replied:
It will error out if the directory is immutable because it will fail to create a file, right?

HastD replied:
Oh I see what you're saying. Yes, I'll add support for handling the immutable flag on the directory in the case where the target file doesn't exist yet. However, this doesn't affect the temp file because it's created in `/tmp`, not in the target directory.

ShadowSlayer1441 commented on files/system/usr/bin/run0edit:
```diff
@@ -71,21 +115,43 @@ else
     editor="$3"
+    edit_immutable=no
+
+    ask_immutable() {
```
Before asking to proceed with an immutable file flagged file, it should check if it's fs is mounted read only. (Otherwise user could assent to proceed only to get a error later anyway.) Something like this could work, theoretically I think you need to check both directory and file thanks to sim-links, but that's quite the edge case.
```elif [ -n "$(findmnt -no OPTIONS --target "$directory" | grep ro)" ]; then```

Also note none of these tests are privileged, so ideally they should be run as user before polkit authentication to save user time.

HastD replied:
Any test that requires read access to the file has to be run in the privileged script, as the file (or even the directory it's in) might not be readable by the user. Anyway, good idea to add the check for a read-only filesystem, I'll put that in.

HastD committed:
feat(run0edit): update to v0.4.1

* Handle immutable flag on directory.
* Check for read-only filesystems to error out earlier.

HastD committed:
Merge branch 'live' into run0edit-update

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -30,29 +31,72 @@ set -eu
+            print_usage
+            exit
+            ;;
+        --)
```
will this ever be hit since getopts isn't being used?

HastD replied:
Yes, a `--` argument can be used to treat the remaining arguments as positional, even if they start with a `-`. In this case since only one positional argument is accepted, it's just a way to handle the edge case of paths that start with a `-`. Probably not strictly necessary but it seems to be a pretty standard convention for this sort of command-line utility.

RoyalOughtness replied:
makes sense

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
     editor="$3"
+    edit_immutable="no"
+
+    readonly_filesystem() {
```
why is this needed twice? even if we need it to be redefined inside the script, we already exit in this case prior to calling the script

HastD replied:
This handles the case where the file is on a read-only filesystem, but neither the file nor its directory are readable at all by the user—in that case we can't detect that it's read-only until we're root.

RoyalOughtness replied:
Ah I see

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+        else
+            immutable_flag=""
+        fi
+        if [ "$immutable_flag" = i ]; then
```
should `i` be in quotes?

HastD replied:
Doesn't make a difference, it doesn't have any special characters so it'll be treated as a string literal either way...

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+        if command -v lsattr >/dev/null; then
+            immutable_flag=$(lsattr -d -- "$1" 2>/dev/null | cut -d " " -f1 | tr -d "-" || echo "")
+        else
+            immutable_flag=""
```
is this required?

HastD replied:
Oh, good point, if `lsattr` isn't found, it'll just fall back to the `echo ""` path anyway... I'll remove the unnecessary conditional.

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -102,37 +202,73 @@ script='\
+                edit_immutable="yes"
+            fi
+            if [ "$edit_immutable" = "no" ]; then
+                echo "run0edit: $filename is read-only."
```
does `edit_immutable` necessarily mean this? seems like it can also mean that the user opted to decline editing it

HastD replied:
If the user declined removing the immutable flag, then the file is read-only, so this message is correct.

RoyalOughtness replied:
Doh, right

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -102,37 +202,73 @@ script='\
                 echo "run0edit: unable to write temporary file at $tmpfile to $filename"
                 exit 1
+            elif [ "$edit_immutable" = "yes" ] && ! cmp -s -- "$tmpfile" "$filename"; then
+                echo "WARNING: contents of $filename does not match contents of tempfile."
```
why is this needed?

HastD replied:
There's a brief window between when the file is overwritten with its new contents and when the immutable flag is reapplied during which another unprivileged program could edit the file. I can't think of any way to get rid of this race condition entirely, so as a best-effort mitigation, we just compare it to the temporary file contents *after* reapplying the immutable flag to make sure they match.

RoyalOughtness replied:
makes sense

RoyalOughtness replied:
@HastD could we use `flock`?

HastD replied:
Isn't the lock set by `flock` optional, in the sense that programs can simply ignore it if they choose? So I don't think it'd help here.

ShadowSlayer1441 replied:
I said it was optional earlier on discord, but reading the flock man page again, maybe it is enforced:

> Since Linux 5.5, flock() locks are emulated with SMB byte-range
       locks on the entire file.  Similarly to NFS, this means that
       fcntl(2) and flock() locks interact with one another.  Another
       important side-effect is that the **locks are not advisory** anymore:
       any IO on a locked file will always fail with EACCES when done
       from a separate file descriptor.  This difference originates from
       the design of locks in the SMB protocol, which provides mandatory
       locking semantics. 

Never mind, I tested this, it doesn't seem to be true. While this was running I was able to open and save the a edit with no problem. I checked with lsof that this program actually locked the file, but it didn't effect anything I tried.
```
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <sys/file.h>

int main() {
	int fd = open("/var/home/user/flock_test_file", O_RDWR);
	if (fd == -1) {
		printf("fd failed\n");
		return 1;
	}

	if (flock(fd, LOCK_EX) == -1) {
		printf("flock failed\n");
		return 1;
	}

	pause();

	flock(fd, LOCK_UN);
	close(fd);
	return 0;
}
```

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -102,37 +202,73 @@ script='\
+            cp -- "$tmpfile" "$filename"
+            cp_exit_status="$?"
+            set -e
+            if [ "$edit_immutable" = "yes" ]; then
```
this works but is error-prone. It would be better to explicitly tie the change in attrs together in the flow. for example if we had the logic for copying the tempfile in a function like `copy_to_dest`

then we could do something like:

```
if edit_immutable
chattr -i
copy_to_dest
chattr +i
else
copy_to_dest
fi
```

this is a picky nit but for something so critical I'm trying to be as picky as possible

HastD replied:
Good idea, I'll do that (and likewise for the "create new file" case).

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,33 +109,89 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+                edit_immutable="yes"
+            fi
+            if [ "$edit_immutable" = "no" ]; then
+                echo "run0edit: $directory is read-only."
```
same thing here: does edit_immutable necessarily mean this? seems like it can also mean that the user opted to decline editing it

HastD replied:
I did end up changing the message here as part of the control flow improvements based on your suggestions.

HastD committed:
improve control flow & message wording

HastD committed:
refactor how inner script is substituted into main script

No change to script functionality with this commit, but single-quotes no
longer need to be avoided in the inner script since it's wrapped in a
heredoc.

HastD committed:
Merge branch 'live' into run0edit-update

HastD commented:
Just FYI, I refactored the script so that, in the [run0edit repo](https://github.com/HastD/run0edit), the main script and inner (privileged) script are now in separate files and are merged into the final `run0edit` script by a (very simple) Python script, `build.py`.

So, you may find it easier to review the inner script at https://github.com/HastD/run0edit/blob/main/run0edit-inner.sh since it has proper syntax highlighting—the `run0edit` file in this PR is produced simply by substituting the contents of `run0edit-inner.sh` into `run0edit-main.sh` in the appropriate place and applying the patch at `run0edit-secureblue.patch` (which doesn't make any changes to the inner script anyway).

ShadowSlayer1441 commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+}
+
+# If the file does not exist, ensure the directory exists and is writable.
+if [ ! -e "$filename" ]; then
```
Why do we only check the directory of the file if it doesn't exist? Can you edit files within an immutable directory? I thought you couldn't. (I can't check rn.)

Edit: nvm, apparently you can edit files within immutable directories.

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+( set -o pipefail 2> /dev/null ) && set -o pipefail
+set -eu
+# shellcheck disable=SC2016
+PATH=$(command -p env -i sh -c 'echo "$PATH"')
```
what's then purpose of this?

HastD replied:
Same thing as in the main script, resets `PATH` to a hard-coded default value.

RoyalOughtness replied:
why not just PATH=""?

HastD replied:
Because then `/usr/bin` wouldn't be in `PATH` and the shell wouldn't be able to find standard utilities?

RoyalOughtness replied:
we could hardcode it to /usr/bin? do we need anything else

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,204 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+            echo 'File created and immutable flag reapplied to directory.'
         else
-            echo "run0edit: $filename not created"
+            touch -- "$filename"
```
this chunk is duplicated and should be factored

HastD replied:
I split out some chunks of code into separate functions, that should get rid of most of the code duplication and make the control flow easier to follow.

HastD committed:
split out duplicated code into functions

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
 fi
 
+readonly_filesystem() {
+    if [ -r "$1" ] && command -pv findmnt >/dev/null; then
```
this could probably be slimmed down to a one-liner, having a separate check for whether the findmnt command exists is redundant I believe

HastD replied:
Good point, and likewise the `[ -r "$1" ]` check can be dropped as `findmnt` will simply fail if the file isn't readable.

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
-        elif [ ! -w "$directory" ]; then
-            echo "run0edit: $directory is read-only."
-            exit 1
+readonly_filesystem() {
```
having these duplicated still smells off to me, but I'm not sure of a better way. passing it between the outer and inner scripts would be a vulnerability...

HastD replied:
I can't think of a better way either, and I don't think it's a big deal to duplicate a single one-liner function (which is what it'll be now).

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
+    readonly edit_immutable="$2"
+    if [ "$edit_immutable" = 'yes' ]; then
+        chattr -i -- "$chattr_target_path"
+        touch -- "$filename"
```
im not sure where the previous comment went but there's still code duplication here. we can move this to:

```bash
copy_to_dest() {
       touch -- "$filename"
        set +e
        cp -- "$tmpfile" "$filename"
        cp_exit_status="$?"
        set -e
}
```

and rename copy_to_dest to copy_immutable_handler, with:

```bash
copy_immutable_handler() {
    readonly chattr_target_path="$1"
    readonly edit_immutable="$2"
    if [ "$edit_immutable" = 'yes' ]; then
        chattr -i -- "$chattr_target_path"
        copy_to_dest
        chattr +i -- "$chattr_target_path"
        echo 'Immutable flag reapplied.'
    else
        copy_to_dest
    fi
...
}
```

RoyalOughtness replied:
the exit status will have to be passed up

HastD replied:
It's just a few lines duplicated here and I think it's clearer this way vs splitting off a second function and passing a value between them, especially given how clunky shell functions are (they don't really have return values). Actually, I think it'd be preferable to just duplicate the `edit_immutable` check, like this:
```sh
    if [ "$edit_immutable" = 'yes' ]; then
        chattr -i -- "$chattr_target_path"
    fi
    touch -- "$filename"
    set +e
    cp -- "$tmpfile" "$filename"
    cp_exit_status="$?"
    set -e
    if [ "$edit_immutable" = 'yes' ]; then
        chattr +i -- "$chattr_target_path"
        echo 'Immutable flag reapplied.'
    fi
```

RoyalOughtness replied:
that's what you had before :smile: but yes I agree that's preferable if we're not using a separate function

my concern still stands though about later down the line someone could inadvertently change the control flow by modifying edit_immutable in-between, but it's a minor concern

HastD replied:
`edit_immutable` is declared as `readonly` though so it'd be an error to try to modify it after initial assignment.

RoyalOughtness replied:
Oh nvm, my concern is mitigated already then

RoyalOughtness commented on files/system/usr/bin/run0edit:
```diff
@@ -65,93 +109,181 @@ elif [ -f '/usr/bin/nano' ] && [ -x '/usr/bin/nano' ]; then
-    tmpfile="$2"
-    editor="$3"
+script=$(cat <<-'END_OF_INNER_SCRIPT'
+#!/bin/sh
```
does this imply that we intend to support any shell? do we want this?

HastD replied:
Yes, I've written the script to work in any POSIX-compatible shell; that's been the case since the first version. My intent is for it to be usable on any Linux system using systemd, so I've avoided using any bash-specific extensions.

HastD committed:
simplify read-only filesystem check, reduce code duplication
